model/scope.go

```diff
@@ -0,0 +1,266 @@
+	PluginSpecificScopePrefix = "pluginscope:"
+)
+
+func ScopeAny(scopes ...Scope) Scopes {
+	return scopes
+}
+
+func ScopeAllow() Scopes {
+	return Scopes{}
+}
+
+func ScopeDeny() Scopes {
+	return nil
+}
+
+func NewPluginScope(pluginID string) Scope {
+	return Scope(PluginScopePrefix + pluginID)
+}
+
+func NewPluginSpecificScope(pluginID, scopeName string) Scope {
+	// This is valid since a plugin can't have a ":" in their ID
+	return Scope(PluginSpecificScopePrefix + pluginID + ":" + scopeName)
+}
+
+func (s Scope) IsPluginScope() bool {
+	return strings.HasPrefix(string(s), PluginScopePrefix)
+}
+
+func (s Scope) IsPluginSpecificScope() bool {
+	return strings.HasPrefix(string(s), PluginSpecificScopePrefix)
+}
+
+func getPredefinedScopes() Scopes {
+	return Scopes{
+		ScopeChannelJoin,
+		ScopeMessageRead,
+		ScopeMessageWrite,
+		ScopeFilesRead,
+		ScopeFilesWrite,
+		ScopeSearchPosts,
+	}
+}
+
+func (s Scope) IsPredefinedScope() bool {
+	return s.isInScope(getPredefinedScopes())
+}
+
+func (s Scope) IsScopeForPlugin(pluginID string) bool {
+	return s.IsPluginScope() &&
+		strings.TrimPrefix(string(s), PluginScopePrefix) == pluginID
+}
+
+// isInScope checks if a scope is in the scope list
+func (s Scope) isInScope(scopes Scopes) bool {
+	if scopes == nil {
+		return false
+	}
+
+	for _, allowed := range scopes {
+		if s == allowed {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (ss Scopes) AreAllowed(allowed Scopes) bool {
+	// To support legacy OAuth Apps, we consider nil scopes as a non-scoped OAuth App.
+	if ss == nil {
+		return true
+	}
+
+	// Allowed endpoints will just relay the scopes, in case there is any consideration to be made based on scopes.
+	if allowed.Equals(ScopeAllow()) {
+		return true
+	}
+
+	if allowed.Equals(ScopeDeny()) {
+		return false
+	}
+
+	if len(ss.intersection(allowed)) == 0 {
+		return false
+	}
+
+	return true
+}
+
+func (ss Scopes) IsPluginInScope(pluginID string) bool {
+	if ss == nil {
+		return true
+	}
+
+	for _, allowed := range ss {
+		if allowed.IsScopeForPlugin(pluginID) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (ss Scopes) intersection(scope Scopes) Scopes {
+	if ss == nil || scope == nil {
+		return ScopeDeny()
+	}
+
+	out := Scopes{}
+	for _, x := range ss {
+		if x.isInScope(scope) {
+			out = append(out, x)
+		}
+	}
+
+	if len(out) == 0 {
+		return ScopeDeny()
+	}
+
+	return out
+}
+
+func (ss Scopes) Validate() bool {
+	for _, s := range ss {
+		if !s.IsPluginScope() &&
+			!s.IsPluginSpecificScope() &&
+			!s.IsPredefinedScope() {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (ss Scopes) Equals(ss2 Scopes) bool {
+	if ss == nil {
+		return ss2 == nil
+	}
+
+	if ss2 == nil {
+		return false
+	}
+
+	return len(ss) == len(ss2) && len(ss) == len(ss.intersection(ss2))
+}
+
+func (ss Scopes) IsSuperset(ss2 Scopes) bool {
+	if ss == nil {
+		return true
+	}
+
+	if ss2 == nil {
+		return false
+	}
+
+	if len(ss2) > len(ss) {
+		return false
+	}
```

Comment:
Are we assuming the "sets" have been deduplicated here? Otherwise this could potentially return wrong results.
Slightly tangential question but did you consider using a map to implement scopes? It seems like a closer data structure to what you need.

Reply:
I did not consider it, no. Usually, they are represented as a list of strings. Not sure if it is worth now making the change to a map.

Reply:
Of course not, that was mostly a curiosity of mine. The only real problem with using arrays (slices) is dealing with deduplication really so as long as you do that properly it should be fine.

Reply:
Since `Scopes` is already a type, can be a map in go, serialized as an array in JSON